File: RemoveDuplicatesFromSortedArrayII.java
public class Solution {
	public int removeDuplicates(int[] A) {
		if(A.length < 3) {
			return A.length;
		}
		HashMap<Integer, Integer> counter = new HashMap<Integer, Integer>();
		int value = A[0];
		counter.put(A[0], 1);
		int index = 1;
		for(int i = 1; i < A.length; i++) {
			if(A[i] == value) {
				if(counter.get(A[i]) == null) {
					counter.put(A[i], 1);
					A[index++] = A[i];
				} else if (counter.get(A[i]) < 2) {
					counter.put(A[i], counter.get(A[i]) + 1);
					A[index++] = A[i];
				} else {
					continue;
				}
			} else {
				A[index++] = A[i];
				value = A[i];
				counter.put(A[i], 1);
			}
		}
		return index;
	}
}

File: RemoveDuplicatesFromSortedArrayII.cc
#include <iostream>
#include <map>

using namespace std;

class Solution {
public:
    int removeDuplicates(int A[], int n) {
        if (n < 3) {
            return n;
        }
        map<int, int> number_counter;
        int index = 0, new_size = 0;
        while(index < n) {
            if (number_counter.find(A[index]) == number_counter.end()) {
                number_counter[A[index]] = 1;
                A[new_size++] = A[index++];
            } else if (number_counter[A[index]] < 2) {
                number_counter[A[index]]++;
                A[new_size++] = A[index++];
            } else {
                index++;
            }
        }
        return new_size;
    }
};

int main()
{
    Solution sol;
    int input[3] = {1, 1, 1};
    cout << sol.removeDuplicates(input, 3) << endl;
}
